src/modules/nfts/components/wizard/MintInProgress.tsx

```diff
@@ -0,0 +1,334 @@
+    }
+
+    try {
+      if (!process.env.NEXT_PUBLIC_SOLANA_ENDPOINT) {
+        throw new Error('No Solana Endpoint');
+      }
+      const metaProgramId = new PublicKey(META_PROGRAM_ID);
+      const { metadata } = mintResp;
+
+      await holaSignMetadata({
+        solanaEndpoint: process.env.NEXT_PUBLIC_SOLANA_ENDPOINT,
+        metadata,
+        metaProgramId,
+        onProgress: (status: signMetaDataStatus) => {
+          console.log('progress status: ', status);
+        },
+        onComplete: () => {
+          console.log('signing complete');
+          setTransactionStep(TransactionStep.SUCCESS);
+          handleNext();
+        },
+        onError: (msg) => {
+          throw new Error(msg);
+        },
+      });
+    } catch (err) {
+      console.log('signing err', err);
+      setTransactionStep(TransactionStep.SIGNING_FAILED);
+    }
+  }, [setTransactionStep, mintResp, handleNext]);
+
+  const attemptMint = useCallback(
+    async (metaData: UploadedFilePin) => {
+      if (!metaData) {
+        throw new Error('No Meta Data, something went wrong');
+      }
+
+      setTransactionStep(TransactionStep.APPROVING);
+
+      try {
+        const mintResp = await mintNFT({
+          connection,
+          wallet,
+          uri: metaData.uri,
+          maxSupply: nftValue.properties.maxSupply,
+        });
+        setMintResp(mintResp);
+        setTransactionStep(TransactionStep.FINALIZING);
+        await connection.confirmTransaction(mintResp.txId);
+        setTransactionStep(TransactionStep.SUCCESS);
+      } catch (err) {
+        if (err?.code === APPROVAL_FAILED_CODE) {
+          setTransactionStep(TransactionStep.APPROVAL_FAILED);
+        } else {
+          setTransactionStep(TransactionStep.SENDING_FAILED);
+        }
+
+        return;
+      }
+
+      setTransactionStep(TransactionStep.SIGNING);
+    },
+    [nftValue, connection, wallet]
+  );
+
+  const attemptMetaDataUpload = useCallback(async () => {
+    if (!nftValue) {
+      throw new Error('No NFT Value, something went wrong');
+    }
+
+    if (nftValue.mintStatus === MintStatus.SUCCESS) {
+      return; // Don't accidentally mint
+    }
+
+    let metaData: UploadedFilePin | null = null;
+
+    try {
+      metaData = await uploadMetaData(nftValue);
+
+      attemptMint(metaData);
+    } catch (err) {
+      setTransactionStep(TransactionStep.META_DATA_UPLOAD_FAILED);
+    }
+  }, [attemptMint, nftValue, uploadMetaData, setTransactionStep]);
+
+  useEffect(() => {
+    if (showErrors) {
+      return;
+    }
+
+    if (isActive && nftValue) {
+      if (transactionStep === TransactionStep.META_DATA_UPLOADING) {
+        attemptMetaDataUpload();
+      } else if (transactionStep === TransactionStep.SIGNING) {
+        attemptHolaplexSign();
+      }
+    }
+  }, [
+    isActive,
+    showErrors,
+    nftValue,
+    transactionStep,
+    attemptHolaplexSign,
+    attemptMetaDataUpload,
+    index,
+  ]);
```

Comment:
This spot is interesting. Can someone break down what is happening here?

We are triggering actions based on state changes?

Reply:
So at some point I realized it easier to just check the status and call the necessary function based on the Enum rather than call the function itself.  Especially with the retries and needing to fire off immediately after the component had mounted.  If it's confusing, I'd be happy to look at a simpler refactor but it does work seemingly pretty smooth.  Maybe something to combine if we bring in Antd Steps.